lib/zclient.h

```diff
@@ -611,6 +625,22 @@ struct zclient_options {
 
 extern struct zclient_options zclient_options_default;
 
+/*
+ * We reserve the top 4 bits for l2-NHG, everything else
+ * is for zebra/proto l3-NHG.
+ *
+ * Each client is going to get it's own nexthop group space
+ * and we'll separate them, we'll figure out where to start based upon
+ * the route_types.h
+ */
+#define ZEBRA_NHG_PROTO_UPPER                                                  \
+	((uint32_t)250000000) /* Bottom 28 bits then rounded down */
+#define ZEBRA_NHG_PROTO_SPACING (ZEBRA_NHG_PROTO_UPPER / ZEBRA_ROUTE_MAX)
+#define ZEBRA_NHG_PROTO_LOWER                                                  \
+	(ZEBRA_NHG_PROTO_SPACING * (ZEBRA_ROUTE_CONNECT + 1))
+
+extern uint32_t zclient_get_nhg_start(uint32_t proto);
```

Comment:
So the idea is that each daemon gets a chunk of number-space to use - that's so that zebra can just use the daemons' numbers directly, I guess? But the api here never tells the daemon the actual range that's available to it? I wonder - why not just have zebra manage the 'shared' number space - the daemon could send whatever number it wanted, and zebra could manage the mapping between the 'owning' daemon's number and the shared/kernel/fib number?

Reply:
I dont think I agree with the conversion. There is around 8million nexthops allocated per protocol which is plenty.

I'd rather have the ID bgp thinks its using the be same one being used by zebra. And its up to bgp to manage how it uses that ID space.

> But the api here never tells the daemon the actual range that's available to it
Zebra tells daemons their starting ID.

Given the daemons their own ID space significantly reduces complexity in zebra. And we dont have to worry about negotiating an ID or races on allocating one between daemons.

With these kind of nexthops, zebra really just makes sure they remain in the kernel and its database and does any replace operations. Zebra doesn't do much beyond that. Other daemons cant hash to them and use them so no issues with a proto wanting to remove/change one another is using.